File: app/src/main/java/ch/unstable/ost/api/model/PassingCheckpoint.java
package ch.unstable.ost.api.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import com.google.common.base.Objects;

import java.util.Date;

import ch.unstable.ost.utils.ParcelUtils;

import static com.google.common.base.Preconditions.checkNotNull;

public class PassingCheckpoint extends Checkpoint implements Parcelable {

    public static final Creator<PassingCheckpoint> CREATOR = new Creator<PassingCheckpoint>() {
        @Override
        public PassingCheckpoint createFromParcel(Parcel in) {
            return new PassingCheckpoint(in);
        }

        @Override
        public PassingCheckpoint[] newArray(int size) {
            return new PassingCheckpoint[size];
        }
    };

    private final Date departureTime;
    private final Date arrivalTime;

    public PassingCheckpoint(Date arrivalTime, Date departureTime, Location location, String platform) {
        super(platform, location);
        this.arrivalTime = checkNotNull(arrivalTime, "arrivalTime");
        this.departureTime = checkNotNull(departureTime, "departureTime");
    }

    private PassingCheckpoint(Parcel in) {
        super(in);
        departureTime = ParcelUtils.readDate(in);
        arrivalTime = ParcelUtils.readDate(in);
    }

    @NonNull
    @Override
    public Date getDisplayDate() {
        return departureTime;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        ParcelUtils.writeDate(dest, departureTime);
        ParcelUtils.writeDate(dest, arrivalTime);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public Date getDepartureTime() {
        return departureTime;
    }

    public Date getArrivalTime() {
        return arrivalTime;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        PassingCheckpoint that = (PassingCheckpoint) o;
        return Objects.equal(departureTime, that.departureTime) &&
                Objects.equal(arrivalTime, that.arrivalTime);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(super.hashCode(), departureTime, arrivalTime);
    }
}

File: app/src/main/java/ch/unstable/ost/api/model/PassingCheckpoint.kt
package ch.unstable.ost.api.model

import android.os.Parcel
import android.os.Parcelable

import com.google.common.base.Objects

import java.util.Date

import ch.unstable.ost.utils.ParcelUtils

data class PassingCheckpoint(val arrivalTime: Date?, val departureTime: Date?, val stopLocation: StopLocation) : Parcelable {

    /**
     * Construct a new passing checkpoint
     *
     * if [arrivalTime] and [departureTime] is null it is considered to be a
     * checkpoint that doesn't stop
     *
     * TODO: 26.09.17  TBD: if arrivalTime or departureTime is null is boarding allowed?
     *
     * @param arrivalTime the arrival time
     * @param departureTime the departure time
     * @param location the location
     * @param platform the platform
     */
    constructor(arrivalTime: Date?, departureTime: Date?, location: Location, platform: String?) : this(arrivalTime, departureTime, StopLocation(platform, location))

    private constructor(source: Parcel) :this(
            departureTime = ParcelUtils.readDate(source),
            arrivalTime = ParcelUtils.readDate(source),
            stopLocation = ParcelUtils.readParcelable(source, StopLocation.CREATOR)!!)

    override fun writeToParcel(dest: Parcel, flags: Int) {
        ParcelUtils.writeDate(dest, departureTime)
        ParcelUtils.writeDate(dest, arrivalTime)
        ParcelUtils.writeParcelable(dest, stopLocation, flags);
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR: Parcelable.Creator<PassingCheckpoint> {
        override fun createFromParcel(`in`: Parcel): PassingCheckpoint {
            return PassingCheckpoint(`in`)
        }

        override fun newArray(size: Int): Array<PassingCheckpoint?> {
            return arrayOfNulls(size)
        }
    }

    val stationName: String
        get() = stopLocation.location.name

    val location: Location
        get() = stopLocation.location

    val platform: String?
        get() = stopLocation.platform
}
